### PR TITLE
List OSGI plugins in the export list

### PR DESCRIPTION
**Pull request by alshamams**
The commit fixes OSGi plugins (that have been created using standard OSGi framework with 'Generate OSGi metadata automatically' feature) not being displayed in the export dialog box. This happens because an OSGi project has a pde.bnd file, unlike the eclipse plugins, which has build.properties. Therefore the previous behaviour was to check for the presence of a build.properties file for validating a plugin project. This assumption breaks with the new type.

The commit relies on the following assumptions in case of OSGi template projects:
 - The location of the pde.bnd file is always under the root folder of the plugin project
 - The filename is always constant, namely, pde.bnd.
 
Fixes: #600

### CODE REVIEW

**Comment by gireeshpunathil** [comment]
the second assumption (instruction file name is always `pde.bnd`) is definitely right due to:
https://github.com/eclipse-pde/eclipse.pde/blob/27cf508efead9b8bfbb8bef99b232f4a3d8ee453/ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/natures/BndProject.java#L24-L26

the first assumption (the project structure) looks to be correct too, but pinging @laeubi to confirm just in case.

**Comment by laeubi** [comment]
Yes `pde.bnd` in the project root is a "marker file" that should always be constant.

**Review by gireeshpunathil — APPROVED** [review]

File: ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/wizards/exports/PluginExportWizardPage.java [hunk 1]
```diff
@@ -71,6 +72,16 @@ private boolean hasBuildProperties(IPluginModelBase model) {
 		return file.exists();
 	}
 
+	private boolean hasPdeBndDescriptor(IPluginModelBase model) {
+		File bin = new File(model.getInstallLocation());
+		if (bin.exists()) {
+			File file = new File(bin.getParent(), ICoreConstants.BND_FILENAME_DESCRIPTOR);
+			return file.exists();
+		}
+		return false;
```
Comment (HannesWell) [inline ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/wizards/exports/PluginExportWizardPage.java:81]: Wouldn't it be better to operate on `IResources` here?
I.e. check if `model.getUnderlyingResource()` returns non null and then check for the `pde.bnd` file of that resource's project?

Comment (HannesWell) [reply ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/wizards/exports/PluginExportWizardPage.java:81]: Furthermore it is actually not necessary to check for the existance of the directory and file.
`File`, just like `Path` or `IFile/IFolder/...` are just handles/pointers and one can create them regardless of if they exist.
Just checking if the finally constructed `(I)File` is an existing file saves one call to the file-system (given the file exists).

Comment (laeubi) [reply ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/wizards/exports/PluginExportWizardPage.java:81]: If you have a project handle one can also use `org.eclipse.pde.internal.core.natures.BndProject.isBndProject(IProject)` (that will only validate the nature) or `org.eclipse.pde.internal.core.bnd.BndProjectManager.getBndProject(IProject)` if also want to validate is has a *valid* instruction file.

Comment (laeubi) [reply ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/wizards/exports/PluginExportWizardPage.java:81]: > Wouldn't it be better to operate on IResources here?
I.e. check if model.getUnderlyingResource() returns non null and then check for the pde.bnd file of that resource's project?

As far as I can see on the callerside the project is already known!

Comment (alshamams) [reply ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/wizards/exports/PluginExportWizardPage.java]: > Furthermore it is actually not necessary to check for the existance of the directory and file. `File`, just like `Path` or `IFile/IFolder/...` are just handles/pointers and one can create them regardless of if they exist. Just checking if the finally constructed `(I)File` is an existing file saves one call to the file-system (given the file exists).

I have implemented the changes suggested by @laeubi , which has made these changes unnecessary, and also doesn't attempt file access.

Comment (alshamams) [reply ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/wizards/exports/PluginExportWizardPage.java]: > If you have a project handle one can also use `org.eclipse.pde.internal.core.natures.BndProject.isBndProject(IProject)` (that will only validate the nature) or `org.eclipse.pde.internal.core.bnd.BndProjectManager.getBndProject(IProject)` if also want to validate is has a _valid_ instruction file.

I have implemented the changes, kindly look through and let me know @laeubi .

File: ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/ICoreConstants.java [hunk 2]
```diff
@@ -60,6 +60,9 @@ public interface ICoreConstants {
 	/** Constant for the string <code>build.properties</code> */
 	String BUILD_FILENAME_DESCRIPTOR = "build.properties"; //$NON-NLS-1$
 
+	/** Constant for the string <code>pde.bnd</code> */
+	String BND_FILENAME_DESCRIPTOR = "pde.bnd"; //$NON-NLS-1$
```
Comment (laeubi) [inline ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/ICoreConstants.java:64]: There is already `org.eclipse.pde.internal.core.natures.BndProject.INSTRUCTIONS_FILE` so you should at laest reference it here
```suggestion
	String BND_FILENAME_DESCRIPTOR = BndProject.INSTRUCTIONS_FILE;
```

File: ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/ICoreConstants.java [hunk 3]
```diff
@@ -60,6 +60,9 @@ public interface ICoreConstants {
 	/** Constant for the string <code>build.properties</code> */
 	String BUILD_FILENAME_DESCRIPTOR = "build.properties"; //$NON-NLS-1$
 
+	/** Constant for the string <code>pde.bnd</code> */
```
Comment (laeubi) [inline ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/ICoreConstants.java:63]: This seems not used anymore in your PR maybe revert this change?

**Comment by alshamams** [comment]
> @alshamams can you please rebase/squash and cleanup the mentioned no longer needed part? The I think this can be merged.

Hi @laeubi ,
I have reverted the change and rebased after squashing.

**Review by laeubi — APPROVED** [review]
For me it looks all sane @HannesWell  @gireeshpunathil any concerns to merge?

**Review by HannesWell — APPROVED** [review]
No concern from me, just a minor, Not required suggestion.

**Comment by HannesWell** [comment]
Thanks @alshamams please squash all commits into one, then we can submit this.

File: ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/wizards/exports/PluginExportWizardPage.java [hunk 4]
```diff
@@ -71,6 +73,15 @@ private boolean hasBuildProperties(IPluginModelBase model) {
 		return file.exists();
 	}
 
+	private boolean hasPdeBndDescriptor(IProject project) {
+		try {
+			return !BndProjectManager.getBndProject(project).isEmpty();
```
[the remark this answers is not in the document]
Comment (alshamams) [reply ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/wizards/exports/PluginExportWizardPage.java]: I have made the changes @HannesWell .

**Comment by alshamams** [comment]
> Thanks @alshamams please squash all commits into one, then we can submit this.

Done @HannesWell

**Review by HannesWell — APPROVED** [review]
Thanks @alshamams.

In general you can actually always squash locally or just amend existing commits and force push as one commit (unless you eventually want to make a change with multiple commits that are self-contained by themself).
After force-push you can still see in the Github UI what as changed by clicking the `Compare` button in the feed:

grafik

Since Gireesh and Christoph were fine with this change too, I'll submit it as soon as the build completes.
